cookbook-react/src/recipes/inputs/selectable-list/components/SelectableList/index.test.tsx

```diff
@@ -0,0 +1,154 @@
+    const radioButton = screen.getByLabelText('First item') as HTMLInputElement;
+    userEvent.click(radioButton);
+    expect(radioButton.checked).toBe(true);
+  });
+
+  test('selects an element when user clicks on its label', () => {
+    const label = screen.getByText('First item') as HTMLInputElement;
+    const radioButton = screen.getByLabelText('First item') as HTMLInputElement;
+    userEvent.click(label);
+    expect(radioButton.checked).toBe(true);
+  });
+
+  test('if an element is selected and another is selected afterwards, the previous element deselects', () => {
+    const radio1 = screen.getByLabelText('First item') as HTMLInputElement;
+    const radio2 = screen.getByLabelText('Second item') as HTMLInputElement;
+    userEvent.click(radio1);
+    userEvent.click(radio2);
+    expect(radio1.checked).toBe(false);
+    expect(radio2.checked).toBe(true);
+  });
+
+  test('if an element is selected and users clicks it, it remains selected', () => {
+    const radio = screen.getByLabelText('First item') as HTMLInputElement;
+    userEvent.click(radio);
+    userEvent.click(radio);
+    expect(radio.checked).toBe(true);
+  });
+});
+
+describe('when the list does not allow multiple selection and starts with initialValue', () => {
+  beforeEach(() => {
+    render(
+      <SelectableList items={items} initialValue="3">
+        {({ item, type, selected, onChange }) => (
+          <div key={item.id}>
+            <input id={item.id} type={type} name={item.name} checked={selected} onChange={onChange} />
+            <label htmlFor={item.id}>{item.label}</label>
+          </div>
+        )}
+      </SelectableList>
+    );
+  });
+
+  test('initializes with only the initial value selected', () => {
+    const radioButtons = screen.getAllByRole('radio') as HTMLInputElement[];
+    const selectedRadioButton = screen.getByLabelText('Third item') as HTMLInputElement;
+    const amountSelected = radioButtons.reduce((result, radio) => result + (radio.checked ? 1 : 0), 0);
+    expect(amountSelected).toBe(1);
+    expect(selectedRadioButton.checked).toBe(true);
+  });
+});
+
+describe('when the list allows multiple selection and starts with no initialValues', () => {
+  beforeEach(() => {
+    render(
+      <SelectableList items={items} multiple>
+        {({ item, type, selected, onChange }) => (
+          <div key={item.id}>
+            <input id={item.id} type={type} name={item.name} checked={selected} onChange={onChange} />
+            <label htmlFor={item.id}>{item.label}</label>
+          </div>
+        )}
+      </SelectableList>
+    );
+  });
+
+  test('initializes with no selected elements', () => {
+    const checkboxes = screen.getAllByRole('checkbox') as HTMLInputElement[];
+    checkboxes.forEach(radio => expect(radio.checked).toBe(false));
+  });
+
+  test('selects an element when user clicks on it', () => {
+    const checkbox = screen.getByLabelText('First item') as HTMLInputElement;
```

Comment:
I was trying to think of a way of avoiding this casts and found this:
```ts
function isElementInput<T extends HTMLElement>(element: T): T is HTMLInputElement {
    // Validate that element is actually an input
    return element instanceof HTMLInputElement;
}

// Update your attempted value sets:
const elem = getByLabelText(/username/i);
if (isElementInput(elem)) {
    elem.value = 'chuck';
} else {
    // Handle failure here...
}
```

Maybe we can add a util like this to the cookbook / bootstrap.

source: https://stackoverflow.com/questions/53195381/react-testing-library-with-typescript-set-an-inputs-value.

Reply:
I don't think it's bad to cast in a test, where it's fine if the test breaks because the element I'm getting is not an input. I might agree if this was production code